File: src/main/java/rails/game/action/SellShares.java
package rails.game.action;

import java.io.IOException;
import java.io.ObjectInputStream;

import com.google.common.base.Objects;

import net.sf.rails.game.CompanyManager;
import net.sf.rails.game.PublicCompany;
import net.sf.rails.util.RailsObjects;
import net.sf.rails.util.Util;

/**
 * Rails 2.0: Updated equals and toString methods
 */
public class SellShares extends PossibleAction {

    // Server-side settings
    private String companyName;
    transient private PublicCompany company;
    private int shareUnit;
    private int shareUnits;
    private int share;
    private int price;
    private int number;
    /** Dump flag, indicates to which type of certificates the president's share must be exchanged.<br>
     * 0 = no dump, or dump that does not require any choice of exchange certificates;<br>
     * 1 = exchange against 1-share certificates (usually 10%);<br>
     * 2 = exchange against a 2-share certificate (as can occur in 1835);<br>
     * etc.
     */
    private int presidentExchange = 0;

    // For backwards compatibility only
    private int numberSold = 0;

    public static final long serialVersionUID = 1L;

    public SellShares(PublicCompany company, int shareUnits, int number,
            int price) {
        this (company, shareUnits, number, price, 0);
    }

    public SellShares(PublicCompany company, int shareUnits, int number,
            int price, int presidentExchange) {
        super(null); // not defined by an activity yet
        this.company = company;
        this.shareUnits = shareUnits;
        this.price = price;
        this.number = number;
        this.presidentExchange = presidentExchange;

        companyName = company.getId();
        shareUnit = company.getShareUnit();
        share = shareUnits * shareUnit;
    }

    /**
     * @return Returns the maximumNumber.
     */
    public int getNumber() {
        return number;
    }

    /**
     * @return Returns the price.
     */
    public int getPrice() {
        return price;
    }

    /**
     * @return Returns the companyName.
     */
    public String getCompanyName() {
        return companyName;
    }

    public PublicCompany getCompany() {
        return getCompanyManager().getPublicCompany(companyName);
    }

    public int getShareUnits() {
        return shareUnits;
    }

    public int getShareUnit() {
        return shareUnit;
    }

    public int getShare() {
        return share;
    }

    public int getPresidentExchange() {
        return presidentExchange;
    }

    @Override
    protected boolean equalsAs(PossibleAction pa, boolean asOption) {
        // identity always true
        if (pa == this) return true;
        //  super checks both class identity and super class attributes
        if (!super.equalsAs(pa, asOption)) return false; 

        // check asOption attributes
        SellShares action = (SellShares) pa;

        return Objects.equal(this.company, action.company)
                && Objects.equal(this.shareUnit, action.shareUnit)
                && Objects.equal(this.shareUnits, action.shareUnits)
                && Objects.equal(this.share, action.share)
                && Objects.equal(this.price, action.price)
                && Objects.equal(this.number, action.number)
        //        && Objects.equal(this.presidentExchange, action.presidentExchange)
        ;
        // no asAction attributes to be checked
    }

    @Override
    public String toString() {
        return super.toString() + 
                RailsObjects.stringHelper(this)
                    .addToString("company", company)
                    .addToString("shareUnit", shareUnit)
                    .addToString("shareUnits", shareUnits)
                    .addToString("share", share)
                    .addToString("price", price)
                    .addToString("number", number)
                    .addToString("presidentExchange", presidentExchange)
                    .toString()
        ;
    }

    /** Deserialize */
    private void readObject(ObjectInputStream in) throws IOException,
    ClassNotFoundException {

        //in.defaultReadObject();
        // Custom reading for backwards compatibility
        ObjectInputStream.GetField fields = in.readFields();

        companyName = (String) fields.get("companyName", null);
        shareUnit = fields.get("shareUnit", shareUnit);
        shareUnits = fields.get("shareUnits", shareUnits);
        share = fields.get("share", share);
        price = fields.get("price", price);
        numberSold = fields.get("numberSold", 0); // For backwards compatibility
        number = fields.get("number", numberSold);
        presidentExchange = fields.get("presidentExchange", 0);

        CompanyManager companyManager = getCompanyManager();
        if (Util.hasValue(companyName))
            companyName = companyManager.checkAlias(companyName);
        company = companyManager.getPublicCompany(companyName);
    }
}

File: dotnet/RailsLib.Net/Rails/Game/Action/SellShares.cs
using GameLib.Net.Game;
using GameLib.Net.Util;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Text;

namespace GameLib.Rails.Game.Action
{
    [JsonObject(MemberSerialization.Fields)]
    public class SellShares : PossibleAction
    {
        // Server-side settings
        private string companyName;
        /*transient*/
        [JsonIgnore]
        private PublicCompany company;
        private int shareUnit;
        private int shareUnits;
        private int share;
        private int price;
        private int number;
        /** Dump flag, indicates to which type of certificates the president's share must be exchanged.<br>
         * 0 = no dump, or dump that does not require any choice of exchange certificates;<br>
         * 1 = exchange against 1-share certificates (usually 10%);<br>
         * 2 = exchange against a 2-share certificate (as can occur in 1835);<br>
         * etc.
         */
        private int presidentExchange = 0;

        // For backwards compatibility only
#pragma warning disable 414
        private int numberSold = 0;
#pragma warning restore 414

        new public const long serialVersionUID = 1L;

        public SellShares(PublicCompany company, int shareUnits, int number, int price) :
            this(company, shareUnits, number, price, 0)
        {

        }

        public SellShares(PublicCompany company, int shareUnits, int number, int price, int presidentExchange) : base(null)
        {
            //super(null); // not defined by an activity yet
            this.company = company;
            this.shareUnits = shareUnits;
            this.price = price;
            this.number = number;
            this.presidentExchange = presidentExchange;

            companyName = company.Id;
            shareUnit = company.GetShareUnit();
            share = shareUnits * shareUnit;
        }

        /**
         * @return Returns the maximumNumber.
         */
        public int Number
        {
            get
            {
                return number;
            }
        }

        /**
         * @return Returns the price.
         */
        public int Price
        {
            get
            {
                return price;
            }
        }

        /**
         * @return Returns the companyName.
         */
        public string CompanyName
        {
            get
            {
                return companyName;
            }
        }

        public PublicCompany Company
        {
            get
            {
                return CompanyManager.GetPublicCompany(companyName);
            }
        }

        public int ShareUnits
        {
            get
            {
                return shareUnits;
            }
        }

        public int ShareUnit
        {
            get
            {
                return shareUnit;
            }
        }

        public int Share
        {
            get
            {
                return share;
            }
        }

        public int PresidentExchange
        {
            get
            {
                return presidentExchange;
            }
        }

        override protected bool EqualsAs(PossibleAction pa, bool asOption)
        {
            // identity always true
            if (pa == this) return true;
            //  super checks both class identity and super class attributes
            if (!base.EqualsAs(pa, asOption)) return false;

            // check asOption attributes
            SellShares action = (SellShares)pa;

            return company.Equals(action.company)
                    && shareUnit.Equals(action.shareUnit)
                    && (shareUnits == action.shareUnits)
                    && (share == action.share)
                    && (price == action.price)
                    && (number == action.number);
            //        && Objects.equal(this.presidentExchange, action.presidentExchange)

            // no asAction attributes to be checked
        }

        override public string ToString()
        {
            return base.ToString() +
                    RailsObjects.GetStringHelper(this)
                        .AddToString("company", company)
                        .AddToString("shareUnit", shareUnit)
                        .AddToString("shareUnits", shareUnits)
                        .AddToString("share", share)
                        .AddToString("price", price)
                        .AddToString("number", number)
                        .AddToString("presidentExchange", presidentExchange)
                        .ToString();
        }

        /** Deserialize */
        [OnDeserialized]
        new internal void OnDeserialized(StreamingContext context)
        {
            CompanyManager companyManager = CompanyManager;
            if (!string.IsNullOrEmpty(companyName))
            {
                companyName = companyManager.CheckAlias(companyName);
            }
            company = companyManager.GetPublicCompany(companyName);
        }
    }
}
